### PR TITLE
Fix Null pointer exception while parsing number

### PR DESCRIPTION
**Pull request by vidhya-45**
This PR fixes a null pointer exception by removing commas from numeric strings before parsing.

Changes:
- Modified `parse-number` function in `src/yang/lang.clj` to strip commas from input strings
- This allows parsing numbers with comma separators (e.g., "1,000" -> 1000)

### CODE REVIEW

**Comment by tolitius** [comment]
the problem with this solution it ignores the comma grouping:

```
=> (parse-number "7,9,6,7.50")
7967.5
```

which is not the right number

we can probably do something like:

```clojure
(defn parse-number [n]
  (let [s (str n)]
    (when (re-matches #"^-?(?:\d+|\d{1,3}(?:,\d{3})+)(?:\.\d+)?$" s)
      (edn/read-string (str/replace s #"," "")))))
```

to respect the grouping

File: src/yang/lang.clj [hunk 1]
```diff
@@ -190,8 +190,8 @@
 
 (defn parse-number [n]
   (let [s (str n)]
-    (if (re-find #"^-?\d+\.?\d*$" s)
-      (edn/read-string s))))
+    (when (re-matches #"^-?(?:\d+|\d{1,3}(?:,\d{3})+)(?:\.\d+)?$" s)
+      (edn/read-string (s/replace s "," "")))))
```
Comment (tolitius) [inline src/yang/lang.clj]: since this is now a large(r) regex
just want to make sure we do not miss a corner case it it fails

this would be a good point to add a test (only for this function for now)
to make sure corner cases are covered

when adding a test file, you cal check [inquery](https://github.com/tolitius/inquery/blob/master/test/inquery/test/core.clj) or [obiwan](https://github.com/tolitius/obiwan/blob/master/test/obiwan/test/search.clj), etc.. on how it is done: namespace, Makefile, etc..

**Review by tolitius — CHANGES_REQUESTED** [review]

File: src/yang/lang.clj (hunk 1 above)
Comment (vidhya-45) [reply src/yang/lang.clj]: Thanks @tolitius, I have added test cases for parse-number function.

While writing tests, I noticed an issue with how numbers with leading zeros are parsed:

**Examples:**
(parse-number "012")    => 10     ; expected 12?
(parse-number "-012")   => -10    ; expected -12?
(parse-number "-012.5") => -10.5  ; expected -12.5?
(parse-number "-05")    => -5     ; expected -5 (correct)
(parse-number "-077")   => -63    ; expected -77?
Root cause: When an input string has leading zeros, edn/read-string interprets it as octal (base-8), not decimal. This follows Clojure's literal parsing rules where integer tokens starting with 0 are treated as octal.


**Solution**
To avoid breaking existing code that may rely on this EDN behavior, I introduced a new function `parse-number-literal `that treats leading zeros as decimal.
parse-number-literal - Decimal-only parsing
This function normalizes leading zeros before parsing:
1. Validates the input against the same regex (comma grouping, format)
2. Removes commas
3. Extracts sign (- if present) and splits integer/fractional parts
4. Strips leading zeros from the integer part (keeps "0" if all zeros)
5. Reassembles the normalized string
6. Parses using Long/valueOf or Double/valueOf (based on decimal point)

**Examples:**
;; parse-number (EDN strict - octal for leading zeros)
(parse-number "045")      => 37    ; octal
;; parse-number-literal (decimal only - normalized)
(parse-number-literal "045")      => 45    ; decimal
(parse-number-literal "-012")     => -12
(parse-number-literal "-012.5")   => -12.5
(parse-number-literal "00.5")     => 0.5
(parse-number-literal "1,234.56") => 1234.56
(parse-number-literal "0x2A")     => nil   ; hex rejected

**Key difference:** By normalizing the integer part before parsing, parse-number-literal avoids EDN's octal interpretation while maintaining the same regex validation and comma-handling as parse-number.

Comment (tolitius) [reply src/yang/lang.clj]: a couple of things

1. check the way I asked to do the test. I gave examples on how the namespaces should be structured
2. "012" is not the right number. "parse-number" is not a custom UI / user input transformer, but a number parser: it takes a string with VALID number, and returns a number
3. the proposed code with all the nesting is difficult to reason about and maintain. AI is a good helper, not yet a good software engineer

this is what we can do (example from the REPL):

```clj
(require '[clojure.string :as str]
         '[clojure.edn :as edn])

(def ^:private number-re
  #"^[+-]?(?:(?:0|[1-9]\d{0,2}(?:,\d{3})+|[1-9]\d*)(?:\.\d+)?|\.\d+)$")

(defn- remove-commas [^String s]
  (if (= -1 (.indexOf s ","))
    s
    (.replace s "," "")))

(defn- valid-number [^String s]
  (when (re-matches number-re s)
    s))

(defn- normalize-leading-decimal [^String s]
  (cond
    (str/starts-with? s ".")  (str "0" s)
    (str/starts-with? s "-.") (str "-0" (subs s 1))
    (str/starts-with? s "+.") (str "+0" (subs s 1))
    :else s))

(defn- parse-it [^String s]
  (edn/read-string s))

(defn parse-number [input]
  (when (string? input)
    (some-> input
            str/trim
            valid-number
            remove-commas
            normalize-leading-decimal
            parse-it)))
```


```clj
=> (parse-number "42")
42
=> (parse-number "42.0")
42.0
=> (parse-number "042.0")
nil
=> (parse-number "42,28.0")
nil
=> (parse-number "42,280.0")
42280.0

user=> (parse-number ".5")
0.5
user=> (parse-number "-.5")
-0.5
user=> (parse-number "+0.523")
0.523
user=> (parse-number "2,3,4")
nil

user=> (parse-number "999999999999999999999999")
999999999999999999999999N
user=> (parse-number "9999999999999999999999.99")
1.0E22
user=> (parse-number "4242424242424242.424242")
4.2424242424242425E15
```

**Review by tolitius — CHANGES_REQUESTED** [review]

File: test/yang/test/lang.clj [hunk 2]
```diff
@@ -0,0 +1,126 @@
+(ns yang.test.lang
```
Comment (tolitius) [inline test/yang/test/lang.clj:1]: better
but almost there

Image

"`make test`" should work as well

**Review by tolitius — CHANGES_REQUESTED** [review]

**Comment by vidhya-45** [comment]
updated **deps.edn** and **Makefile**. 
**`make test`** command works fine (below screenshot)

image

File: test/yang/lang_test.clj [hunk 3]
```diff
@@ -0,0 +1,126 @@
+(ns yang.lang-test
```
Comment (tolitius) [inline test/yang/lang_test.clj]: nope, "`yang.test.lang`" was correct
this is why I gave examples of inquery and obiwan

File: src/yang/lang.clj [hunk 4]
```diff
@@ -188,10 +188,58 @@
...
+ Examples:
+   (parse-number \"42\")       => 42
+   (parse-number \"1,234.56\") => 1234.56
+   (parse-number \".5\")        => 0.5
+   (parse-number \"abc\")      => nil"
+
+  [input]
+  (let [s (cond
+            (string? input) (s/trim input)
+            (number? input) (str input)
```
Comment (tolitius) [inline src/yang/lang.clj]: if it is already a number, we just return it
no need to parse anything

**Review by tolitius — CHANGES_REQUESTED** [review]

File: src/yang/lang.clj [hunk 5]
```diff
@@ -188,10 +188,61 @@
...
+  (parse-number \"42\")       => 42
+  (parse-number \"1,234.56\") => 1234.56
+  (parse-number \".5\")        => 0.5
+  (parse-number \"abc\")      => nil"
+  [input]
+  (cond
+    (and (number? input)
+         (not (Double/isNaN input))
+         (not (Double/isInfinite input)))
+    input
```
Comment (tolitius) [inline src/yang/lang.clj]: no need to use AI here
if it is a number, we return it
the "`(not (Double...`" hallucinations are not needed here

**Review by tolitius — CHANGES_REQUESTED** [review]

File: src/yang/lang.clj [hunk 6]
```diff
@@ -188,10 +188,57 @@
...
+  [input]
+  (cond
+    (number? input) input
+    (string? input) (some-> input
+                            s/trim
+                            valid-number
+                            remove-commas
+                            normalize-leading-decimal
+                            parse-it)
+    :else nil))
```
Comment (tolitius) [inline src/yang/lang.clj]: most of the values passed into this function will be strings
if you use cond, it is going to return after the first matched expression
so it is best to start with "`(string? ..`"

**Review by tolitius — CHANGES_REQUESTED** [review]

**Comment by tolitius** [comment]
is in [[tolitius/yang "0.1.53"]](https://clojars.org/tolitius/yang/versions/0.1.53)

thank you!
